### ISSUE TITLE
Pyright 1.1.256 Not Associating total_ordering decorated class as member of isinstance check.

### ISSUE THREAD

**Issue by bwildenborg** [microsoft/pyright#3633]
**Describe the bug**
Beginning with the latest release of Pyright (1.1.256), it looks like the isinstance check is no longer correctly identifying the type of the object when used with functools.total_ordering.
Back-leveling to Pyright 1.1.255 will resolve the issue, as will removing the total_ordering decorator from the class.

**To Reproduce**
Run the code below on Pyright 1.1.256 command line.
It results in the following errors:
1.   Cannot access member "value" for type "object"  Member "value" is unknown (reportGeneralTypeIssues)
2.   Type of "value" is unknown (reportUnknownMemberType)
3.   Return type is unknown (reportUnknownVariableType)
4.   Cannot access member "value" for type "object"

**Expected behavior**
The isinstance check should correctly associate the object as a member of the checked type.

**Screenshots or Code**
```python
import functools

@functools.total_ordering
class Foo:
    def __init__(self) -> None:
        self.value: int = 0

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, Foo):
            return NotImplemented

        return self.value == other.value

    def __le__(self, other: object) -> bool:
        if not isinstance(other, Foo):
            return NotImplemented

        return self.value <= other.value
```

**VS Code extension or command-line**
Command-line at version 1.1.256

**Comment by erictraut** [comment]
Thanks for the bug report.

This is an interesting interaction between the `@total_ordering` decorator and the `isinstance` type guard logic. I recently fixed a bug in the type guard logic to not apply type narrowing in cases where the type passed as the second argument is a dynamic `type[x]` because this can represent any subclass of `x`, not `x` itself. In your code sample, it looks like you are passing the class `Foo` itself to `isinstance`, but `Foo` here refers to the output of the `total_ordering` decorator, which is `type[Foo]`.

To demonstrate, this modified code type checks without a problem:
```python
class _Foo:
    def __init__(self) -> None:
        self.value: int = 0

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, _Foo):
            return NotImplemented

        reveal_type(other)

        return self.value == other.value

    def __le__(self, other: object) -> bool:
        if not isinstance(other, _Foo):
            return NotImplemented

        return self.value <= other.value
 
Foo = functools.total_ordering(_Foo)
```

I'll need to think more about the proper fix for this.

**Comment by twoertwein** [comment]
Happy to open a new issue but I feel the following might be the same bug.

The following worked with 1.1.255 but fails with 1.1.256:
```py
import pandas as pd  # using the inline type annotations, not pandas-stubs

assert issubclass(pd.IntervalIndex, pd.Index)

def index_fun() -> pd.Index:
    return pd.Index([0, 1, 2])

index = index_fun()

if isinstance(index, pd.IntervalIndex):
    index.left  # error: Cannot access member "left" for type "Index"
```

**Comment by erictraut** [comment]
Yes, that's the same bug. I can confirm that this snippet type checks correctly with my recent fix.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.257, which I just published. It will also be included in the next release of pylance.
